FirebaseInstallations/Source/Library/InstallationsIDController/FIRInstallationsIDController.m

```diff
@@ -234,6 +263,55 @@ - (instancetype)initWithGoogleAppID:(NSString *)appID
       });
 }
 
+- (FBLPromise<FIRInstallationsItem *> *)handleRegistrationRequestError:(NSError *)error
+                                                          installation:
+                                                              (FIRInstallationsItem *)installation {
+  if ([self doesRegistrationErrorRequireConfigChange:error]) {
+    FIRLogError(kFIRLoggerInstallations, kFIRInstallationsMessageCodeInvalidFirebaseConfiguration,
+                @"Firebase Installation registration failed for app with name: %@, error: %@",
+                self.appName, error);
+
+    FIRInstallationsItem *failedInstallation = [installation copy];
+    [failedInstallation updateWithRegistrationError:error
+                             registrationParameters:[self currentRegistrationParameters]];
+
+    // Save the error and then fail with the API error.
+    return
+        [self.installationsStore saveInstallation:failedInstallation].then(^NSError *(id result) {
```

Comment:
If registration returns with error, why do we still save such installation item?

Reply:
I've added a design doc link explaining the details (sorry I should've added it before). TL;DR: we would like to store the registration parameters and the error received, so we can return the error without sending the API request. We will send the API request again only if the parameters changed.

Reply:
In this case `[self doesRegistrationErrorRequireConfigChange:error]` at line #269 will return `NO` so the error will not be saved and the installation status will stay "unregistered".

Reply:
Got it thanks!